File: week1_decomposition1/2_connected_components/ConnectedComponents.java
import java.util.ArrayList;
import java.util.Scanner;
import java.util.Queue;
import java.util.LinkedList;

public class ConnectedComponents {
    private static int numberOfComponents(ArrayList<Integer>[] adj) {
        int captionOfComponent = 0;
        boolean[] visited = new boolean[adj.length];

        for(int vertex = 0; vertex < adj.length; vertex++) {
            if( !visited[vertex] ) {
                captionOfComponent++;
                visited[vertex] = true;
                
                /** 
                 * Breadth first search on the node
                 * All the strongly connected components of the node will
                 * marked as belonging to that component and
                 * then they will be ignored
                 */
                Queue<Integer> queue = new LinkedList<>();
                queue.add(vertex);
                visited[vertex] = true;
                while( !queue.isEmpty() ) {
                    int current = queue.remove();
                    for( int neighbour : adj[current] ) {
                        if (!visited[neighbour] ) {
                            queue.add(neighbour);
                            visited[neighbour] = true;
                        }
                    }
                }
            }
        }

        return captionOfComponent;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int m = scanner.nextInt();
        ArrayList<Integer>[] adj = (ArrayList<Integer>[])new ArrayList[n];
        for (int i = 0; i < n; i++) {
            adj[i] = new ArrayList<Integer>();
        }
        for (int i = 0; i < m; i++) {
            int x, y;
            x = scanner.nextInt();
            y = scanner.nextInt();
            adj[x - 1].add(y - 1);
            adj[y - 1].add(x - 1);
        }
        System.out.println(numberOfComponents(adj));
    }
}

File: week1_decomposition1/2_connected_components/connected_components.py
#Uses python3

import sys
import queue

def number_of_components(adj):
    captionOfConnectedComponents = 0
    visited = [False] * len(adj)
    
    for vertex in range(0, len(adj)):
        if not visited[vertex]:
            captionOfConnectedComponents = captionOfConnectedComponents + 1
            q = queue.Queue(maxsize = len(adj))
            q.put(vertex)
            visited[vertex] = True
            while not q.empty():
                current = q.get()
                for neighbour in adj[current]:
                    if not visited[neighbour]:
                        q.put(neighbour)
                        visited[neighbour] = True

    return captionOfConnectedComponents

if __name__ == '__main__':
    input = sys.stdin.read()
    data = list(map(int, input.split()))
    n, m = data[0:2]
    data = data[2:]
    edges = list(zip(data[0:(2 * m):2], data[1:(2 * m):2]))
    adj = [[] for _ in range(n)]
    for (a, b) in edges:
        adj[a - 1].append(b - 1)
        adj[b - 1].append(a - 1)
    print(number_of_components(adj))
